mindsdb_native/libs/helpers/locking.py

```diff
@@ -1,34 +1,77 @@
 import os
-
+import sys
+import inspect
+import traceback
 import portalocker
 
 from mindsdb_native.config import CONFIG
 
 
-def learn_lock(name):
-    learn_lock_fn = os.path.join(CONFIG.MINDSDB_STORAGE_PATH,f'{name}_learn.lock')
-    f = open(learn_lock_fn, 'a+')
-    portalocker.lock(f, portalocker.LOCK_EX+portalocker.LOCK_NB)
-    return f
-
-def predict_lock(name):
-    learn_lock_fn = os.path.join(CONFIG.MINDSDB_STORAGE_PATH,f'{name}_learn.lock')
-    f = open(learn_lock_fn, 'a+')
-    portalocker.lock(f, portalocker.LOCK_SH+portalocker.LOCK_NB)
-    return f
-
-def delete_lock(name):
-    learn_lock(name)
-    delete_lock_fn = os.path.join(CONFIG.MINDSDB_STORAGE_PATH,f'{name}_delete.lock')
-    f = open(delete_lock_fn, 'a+')
-    portalocker.lock(f, portalocker.LOCK_EX+portalocker.LOCK_NB)
-    return f
-
-def get_data_lock(name):
-    delete_lock_fn = os.path.join(CONFIG.MINDSDB_STORAGE_PATH,f'{name}_delete.lock')
-    f = open(delete_lock_fn, 'a+')
-    portalocker.lock(f, portalocker.LOCK_SH+portalocker.LOCK_NB)
-    return f
-
-def unlock(file):
-    portalocker.unlock(file)
+def mdb_lock(flags, lock_name, argname=None):
+    """
+    :param lock_name: str, name of the lock
+    :argname: str, name of the positional/keyword argument of
+    the input function that will be concatenated with :lock_name:
+    """
+    def wrapper1(func):
+        def wrapper2(*args, **kwargs):
+            if argname is None:
+                final_lock_name = '{}.lock'.format(lock_name)
+            else:
+                if argname.startswith('self.'):
+                    argval = getattr(args[0], argname.split('.')[1])
+                elif argname in kwargs:
+                    argval = kwargs[argname]
+                else:
+                    index = inspect.getfullargspec(func).args.index(argname)
+                    if index < len(args):
+                        argval = args[index]
+                    else:
+                        raise Exception('argname wasn\'t found in *args/**kwargs')
+                final_lock_name = '{}_{}.lock'.format(lock_name, argval)
+
+            path = os.path.join(CONFIG.MINDSDB_STORAGE_PATH, final_lock_name)
+            f = open(path, 'a+')
+            portalocker.lock(f, flags)
+
+            try:
+                ret = func(*args, **kwargs)
+            except BaseException:
```

Comment:
Why except "BaseException" here only ?

I think this should just be `except` since you want to catch everything, intrerrupts included, so that you can unlock the lock at exit, no matter the cause of existing.

Reply:
`finally` clause exists for this case

Reply:
1. `BaseException` catches all possible exceptions (including `KeyboardInterrupt`), I'm using it over `except:` because it's more explicit.

2. `finally:` executes before `except:`, so the lock is unlocked even if there is a new exception raised under `except:`

Reply:
See link above: finally executes after all `except` clauses:
> If a finally clause is present, the finally clause will execute as the last task before the try statement completes.
This is the exact use-case of finally